WordPress/gutenberg
# Wrap around eslint-plugin-jsx-a11y/anchor-has-content to support createInterpolateElement usage

## What?

Fixes https://github.com/WordPress/gutenberg/issues/21441.

Changes:

- packages/eslint-plugin/configs/jsx-a11y.js: disables the rule 'jsx-a11y/anchor-has-content' in favor of wordpress/a11y-interpolated-anchor, which wraps around the former to allow both checking regular JSX and not failing on interpolated elements.

- a11y-interpolated-anchor.js adds the new wrapper rule. This rule is then enabled in the custom.js rules of the eslint-plugin.

## Why?

When using anchors and createInterpolateElement together, we need to eslint-ignore the line to avoid running into the anchor-has-content error even if the anchor is valid. 
This pull request introduces a wrapper around the rule so that we do not need to constantly ignore it.


## How?

- The new rule uses jsx-a11y/anchor-has-content under the hood to validate non-interpolated anchors (as normal). I considered just writing the rule from scratch, and I can still do that but as a first attempt, I decided it made sense to add a wrapper and not reinvent the linter rules for normal anchors.
- The new rule also does some very basic checks for the potential validity of anchor tags used within _createInterpolateElement_. In particular, it checks that anchors are not empty.
- The trade-off of having any linting at all for the interpolated anchor is that it could give a false sense of safety, but it would still miss issues (see the tests for some examples).


## Testing Instructions





### Testing Instructions for Keyboard

- "Break" an anchor that is created with createInterpolateElement, by making it empty of content (e.g: `<a> </a>`)
    - It should trigger a `@wordpress/a11y-anchor-has-content` linter error that says `Anchors must have content and the content must be accessible by a screen reader. Check the first parameter to createInterpolateElement` 
- Add some empty anchor with JSX like so `<a href=""></a></>`, without createInterpolateElement
    - It should trigger a `@wordpress/a11y-anchor-has-content` linter error that says `Anchors must have content and the content must be accessible by a screen reader`
- Run the tests for the eslint-plugin rules `npm run test:unit --  packages/eslint-plugin/rules/__tests__/jsx-a11y-anchor-has-content.js`. The `a11y-anchor-has-content.js` tests should pass (and all others).
